orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2019 Pivotal, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.expressions.functions;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.TreeTraversingParser;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.jackson.OrcaObjectMapper;
+import com.netflix.spinnaker.orca.pipeline.expressions.ExpressionFunctionProvider;
+import com.netflix.spinnaker.orca.pipeline.model.Execution;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import lombok.Data;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.function.Predicate;
+
+@Component
+public class ServiceKeyExpressionFunctionProvider implements ExpressionFunctionProvider {
+  private static final String CREATE_SERVICE_KEY_STAGE_NAME = "createServiceKey";
+  private static final ObjectMapper objectMapper = OrcaObjectMapper.newInstance();
+
+  @Nullable
+  @Override
+  public String getNamespace() {
+    return null;
+  }
+
+  @NotNull
+  @Override
+  public Collection<FunctionDefinition> getFunctions() {
+    return Collections.singletonList(
+      new FunctionDefinition("serviceKey", Arrays.asList(
+        new FunctionParameter(Execution.class, "execution", "The execution to search for stages within"),
+        new FunctionParameter(String[].class, "ids", "A list of stage name or stage IDs to search")
+      ))
+    );
+  }
+
+  public static Map<String, Object> serviceKey(Execution execution, String... id) {
+    return execution.getStages()
+      .stream()
+      .filter(matchesServiceKeyStage(id))
+      .findFirst()
```

Comment:
can there only be one `createServiceKeyStage`?

Reply:
There may be multiple `createServiceKey` stages.  Each will have its own name, and we expect them to be addressed primarily by name rather than by id.  We have removed the spread operator.

Reply:
got it, so the above should be `"id"` instead of `"ids"` (in the params)